src/plot_api/plot_api.js

```diff
@@ -2811,6 +2811,29 @@ function diffLayout(gd, oldFullLayout, newFullLayout, immutable, transition) {
         return PlotSchema.getLayoutValObject(newFullLayout, parts);
     }
 
+    // Clear out any _inputDomain that's no longer valid
+    for (var key in newFullLayout) {
+        if (!key.startsWith('xaxis') && !key.startsWith('yaxis')) {
+            continue;
+        }
+        if (!oldFullLayout[key]) {
+            continue;
+        }
+        var newDomain = newFullLayout[key].domain;
+        var oldDomain = oldFullLayout[key].domain;
+        var oldInputDomain = oldFullLayout[key]._inputDomain;
+        if (oldFullLayout[key]._inputDomain) {
+            if (newDomain[0] === oldInputDomain[0] && newDomain[1] === oldInputDomain[1]) {
+                // what you're asking for hasn't changed, so let plotly.js start with what it
+                // concluded last time and iterate from there
+                newFullLayout[key].domain = oldFullLayout[key].domain;
+            } else if (newDomain[0] !== oldDomain[0] || newDomain[1] !== oldDomain[1]) {
+                // what you're asking for HAS changed, so clear _inputDomain and let us start from scratch
+                newFullLayout[key]._inputDomain = null;
```

Comment:
I think `nestedProperty(...).set(null)` is [actually equivalent to](https://github.com/plotly/plotly.js/blob/ca6a8868c22968afff2f8bd7d26f839df23b9d15/src/lib/nested_property.js#L128) `delete newFullLayout[key]._inputDomain`

Reply:
@alexcjohnson  Using `delete` may be slower comparing to setting to `null`. No?

Reply:
I was using this syntax for consistency with the other places in the codebase that delete this field, but I don't have a strong opinion on which we use.

Reply:
@marthacryan Maybe worth leaving a comment explaining why we do nothing in the `else` case? (I know we discussed but... for posterity)

Reply:
Ah yeah good point Emily!

Reply:
@alexcjohnson I added a [comment](https://github.com/plotly/plotly.js/pull/7283/commits/5f128bbe7b362dcd4d4a7db47ba43452f5ffbf14) explaining the else case based off of this explanation from you about that case:
"this situation is perhaps actually ambiguous: did you pass back in the previous layout as modified by Plotly.js, so we should assume you really wanted _inputDomain to still apply? or did you explicitly set the new domain and it just happened to match the old final calculated domain - in which case we should clear _inputDomain. I guess generally oldDomain will be some random floating point number so it's unlikely you matched it exactly. better to assume you passed back in the same layout object and change nothing in this case."

Let me know if you have any corrections!